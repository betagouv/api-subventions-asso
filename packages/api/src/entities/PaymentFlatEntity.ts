import * as Sentry from "@sentry/node";
import {
    idEntrepriseType,
    idEtablissementType,
    typeIdEntreprise,
    typeIdEtablissement,
} from "../valueObjects/typeIdentifier";
import { ParserInfo } from "../@types";
import { GenericParser } from "../shared/GenericParser";
import Siret from "../valueObjects/Siret";

export default class PaymentFlatEntity {
    private static regionMapping: Record<string, string> = {
        ADCE: "Administration Centrale",
        DEFE: "Administration Centrale",
        DILA: "Administration Centrale",
        ETR1: "Administration Centrale",

        DOM1: "DOM-TOM",
        DOM2: "DOM-TOM",
        DOM3: "DOM-TOM",
        DOM4: "DOM-TOM",

        ALSA: "Grand Est",
        LORR: "Grand Est",
        CHAR: "Grand Est",

        AQUI: "Nouvelle-Aquitaine",
        LIMO: "Nouvelle-Aquitaine",
        POIT: "Nouvelle-Aquitaine",

        AUVE: "Auvergne-Rhône-Alpes",
        RALP: "Auvergne-Rhône-Alpes",

        BNOR: "Normandie",
        HNOR: "Normandie",

        BOUR: "Bourgogne-Franche-Comté",
        FRCO: "Bourgogne-Franche-Comté",

        BRET: "Bretagne",

        CENT: "Centre-Val de Loire",

        CORS: "Corse",

        IDF1: "Île-de-France",

        LANG: "Occitanie",
        MIPY: "Occitanie",

        NORP: "Hauts-de-France",
        PICA: "Hauts-de-France",

        PACA: "Provence-Alpes-Côte d'Azur",

        PAYL: "Pays de la Loire",
    };

    public static getRegionAttachementComptable(attachementComptable: string | "N/A"): string | "N/A" {
        if (attachementComptable == "N/A") return "N/A";

        const region = PaymentFlatEntity.regionMapping[attachementComptable];
        if (region === undefined) {
            const errorMessage = `Unknown region code: ${attachementComptable}`;
            Sentry.captureException(new Error(errorMessage));
            console.error(errorMessage);
            return "code region inconnu";
        }
        return region;
    }

    public static chorusToPaymentFlatPath: { [key: string]: ParserInfo } = {
        exerciceBudgetaire: {
            path: ["Exercice comptable"],
            adapter: value => {
                if (!value) return value;
                return parseInt(value, 10);
            },
        },
        typeIdEtablissementBeneficiaire: {
            path: ["Code taxe 1"],
            adapter: value => {
                if (Siret.isSiret(value)) return "siret";
            },
        },
        idEtablissementBeneficiaire: {
            path: ["Code taxe 1"],
            adapter: value => {
                if (value) return new Siret(value);
            },
        },
        typeIdEntrepriseBeneficiaire: {
            path: ["Code taxe 1"],
            adapter: value => {
                if (Siret.isSiret(value)) return "siren";
            },
        },
        idEntrepriseBeneficiaire: {
            path: ["Code taxe 1"],
            adapter: value => {
                if (value) return new Siret(value).toSiren();
            },
        },
        amount: {
            path: ["Montant payé"],
            adapter: value => {
                if (!value || typeof value === "number") return value;

                return parseFloat(value.replaceAll(/[\r ]/, "").replace(",", "."));
            },
        },
        dateOperation: {
            path: ["Date de dernière opération sur la DP"],
            adapter: value => {
                if (!value) return value;
                if (value != parseInt(value, 10).toString()) {
                    const [day, month, year] = value.split(/[/.]/).map(v => parseInt(v, 10));
                    return new Date(Date.UTC(year, month - 1, day));
                }
                return GenericParser.ExcelDateToJSDate(parseInt(value, 10));
            },
        },

        centreFinancierCode: {
            path: ["Centre financier CODE"],
        },
        centreFinancierLibelle: {
            path: ["Centre financier"],
        },
        attachementComptable: {
            path: ["Société"],
        },
        ej: {
            path: ["N° EJ"],
        },
        provider: {
            path: [],
            adapter: () => "Chorus",
        },
    };

    public regionAttachementComptable: string | "N/A" | "code region inconnu";
    public idVersement: string;
    public uniqueId: string;

    constructor(
        public exerciceBudgetaire: number,
        public typeIdEtablissementBeneficiaire: typeIdEtablissement,
        public idEtablissementBeneficiaire: idEtablissementType<typeIdEtablissement>,
        public typeIdEntrepriseBeneficiaire: typeIdEntreprise,
        public idEntrepriseBeneficiaire: idEntrepriseType<typeIdEntreprise>,
        public amount: number,
        public operationDate: Date,
        public centreFinancierCode: string | "N/A",
        public centreFinancierLibelle: string | "N/A" | null,
        public attachementComptable: string | "N/A",
        public ej: string,
        public provider: string,
        public programName: string | null,
        public programNumber: number,
        public mission: string | null,
        public ministry: string | null,
        public ministryAcronym: string | null,
        public actionCode: string,
        public actionLabel: string | null,
        public activityCode: string | null,
        public activityLabel: string | null,
    ) {
        this.regionAttachementComptable = PaymentFlatEntity.getRegionAttachementComptable(attachementComptable);
        this.idVersement = `${idEtablissementBeneficiaire}-${ej}-${exerciceBudgetaire}`;
        this.uniqueId = `${
            this.idVersement
        }-${programNumber}-${actionCode}-${activityCode}-${operationDate.getTime()}-${attachementComptable}-${centreFinancierCode}`;
    }
}
