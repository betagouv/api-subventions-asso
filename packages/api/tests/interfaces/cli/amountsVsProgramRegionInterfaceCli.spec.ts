import { ObjectId } from "mongodb";
import amountsVsProgramRegionPort from "../../../src/dataProviders/db/dataViz/amountVSProgramRegion/amountsVsProgramRegion.port";
import paymentFlatPort from "../../../src/dataProviders/db/paymentFlat/paymentFlat.port";
import AmountsVsProgramRegionCli from "../../../src/interfaces/cli/AmountsVsProgramRegion.cli";
import { CHORUS_PAYMENT_FLAT_ENTITY } from "../../../src/modules/paymentFlat/__fixtures__/paymentFlatEntity.fixture";
import PaymentFlatEntity from "../../../src/entities/PaymentFlatEntity";

const MOCK_DOCUMENTS: PaymentFlatEntity[] = [
    // the following two should be aggregated together
    { ...CHORUS_PAYMENT_FLAT_ENTITY, provider: "chorus", uniqueId: "1" },
    { ...CHORUS_PAYMENT_FLAT_ENTITY, amount: 7000, provider: "chorus", uniqueId: "2" },

    // the following two should be aggregated together
    { ...CHORUS_PAYMENT_FLAT_ENTITY, amount: 100_000, programNumber: 1234, provider: "chorus", uniqueId: "3" },
    { ...CHORUS_PAYMENT_FLAT_ENTITY, amount: 30_000, programNumber: 1234, provider: "chorus", uniqueId: "4" },

    // the following two should be aggregated together
    {
        ...CHORUS_PAYMENT_FLAT_ENTITY,
        amount: 34_000,
        attachementComptable: "HNOR",
        regionAttachementComptable: "Normandie",
        exerciceBudgetaire: 2021,
        provider: "chorus",
        uniqueId: "5",
    },
    {
        ...CHORUS_PAYMENT_FLAT_ENTITY,
        amount: 34_000,
        attachementComptable: "BNOR",
        regionAttachementComptable: "Normandie",
        exerciceBudgetaire: 2021,
        provider: "chorus",
        uniqueId: "6",
    },

    // this one should not be aggregated
    {
        ...CHORUS_PAYMENT_FLAT_ENTITY,
        amount: 34_000,
        attachementComptable: "BNOR",
        regionAttachementComptable: "Normandie",
        exerciceBudgetaire: 2024,
        provider: "chorus",
        uniqueId: "7",
    },
];

const insertData = async () => {
    await paymentFlatPort.upsertMany(MOCK_DOCUMENTS);
};

function sortResultForSnapshot(a, b) {
    if (a.regionAttachementComptable < b.regionAttachementComptable) return -1;
    if (a.regionAttachementComptable > b.regionAttachementComptable) return 1;
    if (a.programme < b.programme) return -1;
    if (a.programme > b.programme) return 1;

    return a.exerciceBudgetaire - b.exerciceBudgetaire;
}

describe("AmountsVsProgramRegionCli", () => {
    beforeEach(async () => {
        await amountsVsProgramRegionPort.deleteAll();
        await insertData();
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    const cli = new AmountsVsProgramRegionCli();

    describe("init()", () => {
        it("should persist amounts vs program region collection", async () => {
            await cli.init();
            //@ts-expect-error protected method
            const amountsVsProgramRegion = (await amountsVsProgramRegionPort.collection.find({}).toArray())
                .map(amountsVsProgramRegion => ({
                    ...amountsVsProgramRegion,
                    _id: expect.any(ObjectId),
                }))
                .sort(sortResultForSnapshot);

            expect(amountsVsProgramRegion).toMatchSnapshot("Snapshot init");
        });
    });

    describe("resyncExercice()", () => {
        it("should persist amounts vs program region collection for the given exercice", async () => {
            const exerciceStr = "2023";
            await cli.resyncExercice(exerciceStr);
            //@ts-expect-error protected method
            const amountsVsProgramRegion = (await amountsVsProgramRegionPort.collection.find({}).toArray())
                .map(amountsVsProgramRegion => ({
                    ...amountsVsProgramRegion,
                    _id: expect.any(ObjectId),
                }))
                .sort(sortResultForSnapshot);

            expect(amountsVsProgramRegion).toMatchSnapshot("Snapshot resyncExercice");
        });
    });
});
