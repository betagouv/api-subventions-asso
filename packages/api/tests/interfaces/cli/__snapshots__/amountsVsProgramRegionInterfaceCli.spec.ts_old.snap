// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountsVsProgrammeRegionCli init() should persist amounts vs programme region collection: Snapshot init 1`] = `
Array [
  Object {
    "_id": Any<ObjectId>,
    "exerciceBudgetaire": 2023,
    "mission": "Mission Exemple",
    "montant": 8000,
    "programme": "1 - Programme Exemple",
    "regionAttachementComptable": "Bretagne",
  },
  Object {
    "_id": Any<ObjectId>,
    "exerciceBudgetaire": 2023,
    "mission": "Mission Exemple",
    "montant": 100000,
    "programme": "1234 - Programme Exemple",
    "regionAttachementComptable": "Bretagne",
  },
  Object {
    "_id": Any<ObjectId>,
    "exerciceBudgetaire": 2021,
    "mission": "Mission Exemple",
    "montant": 34000,
    "programme": "1 - Programme Exemple",
    "regionAttachementComptable": "Normandie",
  },
]
`;

exports[`AmountsVsProgrammeRegionCli resyncExercice() should persist amounts vs programme region collection for the given exercice: Snapshot resyncExercice 1`] = `
Array [
  Object {
    "_id": Any<ObjectId>,
    "exerciceBudgetaire": 2023,
    "mission": "Mission Exemple",
    "montant": 8000,
    "programme": "1 - Programme Exemple",
    "regionAttachementComptable": "Bretagne",
  },
  Object {
    "_id": Any<ObjectId>,
    "exerciceBudgetaire": 2023,
    "mission": "Mission Exemple",
    "montant": 100000,
    "programme": "1234 - Programme Exemple",
    "regionAttachementComptable": "Bretagne",
  },
]
`;
